toolchain/parse/handle_match.cpp

```diff
@@ -0,0 +1,232 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/parse/context.h"
+
+namespace Carbon::Parse {
+
+namespace {
+auto HandleStatementsBlockStart(Context& context, State finish,
+                                NodeKind equal_greater, NodeKind starter,
+                                NodeKind complete) -> void {
+  auto state = context.PopState();
+
+  if (!context.PositionIs(Lex::TokenKind::EqualGreater)) {
+    if (!state.has_error) {
+      CARBON_DIAGNOSTIC(ExpectedMatchCaseArrow, Error,
+                        "Expected `=>` introducing statement block.");
+      context.emitter().Emit(*context.position(), ExpectedMatchCaseArrow);
+    }
+
+    context.AddLeafNode(equal_greater, *context.position(), true);
+
+    context.AddNode(starter, *context.position(), state.subtree_start, true);
+
+    context.AddNode(complete, *context.position(), state.subtree_start, true);
+
+    context.SkipPastLikelyEnd(*context.position());
+    return;
+  }
+
+  context.AddLeafNode(equal_greater, context.Consume());
+
+  if (!context.PositionIs(Lex::TokenKind::OpenCurlyBrace)) {
+    if (!state.has_error) {
+      CARBON_DIAGNOSTIC(ExpectedMatchCaseBlock, Error,
+                        "Expected `{{` after `=>`.");
+      context.emitter().Emit(*context.position(), ExpectedMatchCaseBlock);
+    }
+
+    context.AddNode(starter, *context.position(), state.subtree_start, true);
+
+    context.AddNode(complete, *context.position(), state.subtree_start, true);
+
+    context.SkipPastLikelyEnd(*context.position());
```

Comment:
Similarly, compact a bit:

```suggestion
    context.AddNode(starter, *context.position(), state.subtree_start, true);
    context.AddNode(complete, *context.position(), state.subtree_start, true);
    context.SkipPastLikelyEnd(*context.position());
```